File: Problems/876-Middle-of-the-Linked-List/python/solution_1.py
from typing import Optional
from helpers import ListNode


class Solution1:
    """
    Approach: Two Pass Approach or Iterative counting approach
    Time complexity: O(N)
    Space complexity: O(1)

    Logic: Calculate the length of the list, 
    then traverse till the half of the length to find the middle node.
    """

    def middleNode(self, head: Optional[ListNode[int]]) -> Optional[ListNode[int]]:

        length = 0
        curr = head
        # Measure the length of the list
        while curr:
            length += 1
            curr = curr.next
        # Calculate the half length
        length = length // 2
        # Reset current to the start of the list
        curr = head
        # Move forward till we reach the middle
        while curr and length:
            curr = curr.next
            length -= 1
        # Return the middle node
        return curr

File: Problems/876-Middle-of-the-Linked-List/java/Solution1.java
/**
 * Approach: Two Pass Approach or Iterative counting approach
 * Time Complexity: O(N)
 * Space Complexity: O(1)
 */
public class Solution1 {
    ListNode middleNode(ListNode head) {
        int length = 0;
        ListNode curr = head;
        while (curr != null) {
            length += 1;
            curr = curr.next;
        }
        length = Math.floorDiv(length, 2);
        curr = head;
        while (length != 0) {
            curr = curr.next;
            length -= 1;
        }
        return curr;
    }
}
